monitoring/metrics_config.yaml

```diff
@@ -0,0 +1,49 @@
+# This Configuration files details the all metrics Clipper is collecting.
+
+# Four Metric Type {Counter | Gauge | Histogram | Summary},
+# see: https://prometheus.io/docs/concepts/metric_types/
+# Note: Favor Histogram over Summary for its efficiency in client side.
+#       Summary data model requires computation within the exporter.
+#       While Histogram handles the computation within Prometheus server.
+
+Model Container:
+  prefix: mc
+
+  pred_total:
+    type: Counter
+    description: The number of prediction made in this model container
+
+  end_to_end_latency_us:
+    type: Histogram
+    description: The time in ms takes from receive RPC call to send off prediction.
+    bucket: [0, 500, 5] # This will be interpreted as np.linspace(0, 500, 5) + inf
+
+  recv_time_us:
+    type: Histogram
+    description: The time in ms takes to receive RPC call.
+    bucket: [0, 100, 5]
```

Comment:
@dcrankshaw 

That's too many buckets though. Each bucket is a prometheus time series. We should have around 5 to 10 "bucket" for each histogram metric. 

For the simple query example running on my local machine, recv time takes around `40 us`, parse time takes around `100 us`, handle time takes around `90 us`. So the end to end is about `230 us`. 

For the range specification, it works like `[start, end, num_of_interval]`.

How about the following bucket (in us)? I make the general histogram to have a long tail to observe abnormal event.

Type | Time
-|-
recv| `[20, 120, 5]`
parse | `[50, 250, 5]`
handle | `[50, 250, 5]`
end-to-end | `[100, 600, 5]`

In the end, we need to let the user configure it and pass it dynamically somehow. Or configure it according to the user's SLO objective, but that is dynamically configured still. Right now, the metric and bucket is set when building the RPC container, I should think of a better way to dynamically pass it in somehow.

Reply:
Okay that makes sense. Those buckets are fine for recv and parse. But handle will be much longer for any real application. I agree that it makes sense to dynamically set the handle buckets based on the SLO, but for now can we do logarithmically spaced buckets from 1ms to 500 ms. Something like (in ms):
[5, 10, 20, 35, 50, 75, 100, 150, 200, 250, 300, 400, 500]

That's a little bit more than 10 buckets, but for any given application most of those buckets will be empty.

Reply:
in `us`, right? 

Since the model container log is printing:
```
Got start of message 145
recv: 62.000000 us, parse: 100.000000 us, handle: 81.000000 us
...
```

Reply:
No in ms. Are you testing with the noop model container? Any real models will take substantially longer. E.g. a deep learning model may take up to a few hundred ms.

Actually now that I think about it let's change the recv and parse histograms to use (in us) `[100, 1000, 100]` (so go up to 1 ms in 100 us increments). We don't need fine-grained latency tracking on the order of microseconds or tens of microseconds. Remember that some applications may be sending very larger inputs that take much longer than a few microseconds to recv and parse.

Reply:
Ok what about the following:

```yaml
Model Container:
  prefix: mc

  pred_total:
    type: Counter
    description: The number of prediction made in this model container

  end_to_end_latency_us:
    type: Histogram
    description: The time in us takes from receive RPC call to send off prediction.
    bucket: 
      log: [100, 10000, 10]
          #[  100,   170,   280,   460,   770,  1290,  2150,  3590,  5990, 10000]
  recv_time_us:
    type: Histogram
    description: The time in us takes to receive RPC call.
    bucket: 
      linear: [100, 1000, 10]
                #[ 100,  200,  300,  400,  500,  600,  700,  800,  900, 1000]
  parse_time_us:
    type: Histogram
    description: The time in us takes to parse RPC call.
    bucket: 
      linear: [100, 1000, 10]
                #[ 100,  200,  300,  400,  500,  600,  700,  800,  900, 1000]
  handle_time_us:
    type: Histogram
    description: The time in us takes to make prediction.
    bucket: 
      log: [100, 10000, 10]
           #[  100,   170,   280,   460,   770,  1290,  2150,  3590,  5990, 10000]
```

Reply:
10 ms for handle time and end-to-end time is not enough. Handle time is the time for the model to make a prediction. It needs to go up to 500 ms. Can you just explicitly set the bucket boundaries to what I specified before ([5, 10, 20, 35, 50, 75, 100, 150, 200, 250, 300, 400, 500])?

Reply:
Sure! The implementation in RPC will be simpler too! 

```yaml
Model Container:
  prefix: mc

  pred_total:
    type: Counter
    description: The number of prediction made in this model container

  end_to_end_latency_us:
    type: Histogram
    description: The time in us takes from receive RPC call to send off prediction.
    bucket: [5, 10, 20, 35, 50, 75, 100, 150, 200, 250, 300, 400, 500]

  recv_time_us:
    type: Histogram
    description: The time in us takes to receive RPC call.
    bucket: [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]

  parse_time_us:
    type: Histogram
    description: The time in us takes to parse RPC call.
    bucket: [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]

  handle_time_us:
    type: Histogram
    description: The time in us takes to make prediction.
    bucket: [5, 10, 20, 35, 50, 75, 100, 150, 200, 250, 300, 400, 500]

```

Reply:
Great. Yeah this looks good. It's probably better to use the same time units for all the histograms though (this will avoid future confusion). Can you make all of the microseconds into milliseconds and just use floats for recv and parse time?

Reply:
Sure! Done.